### PR TITLE
[Form] Add Bootstrap 4 style for field FileType

### PR DESCRIPTION
**Pull request by zzenmate**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #26464   
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

### CODE REVIEW

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 1]
```diff
@@ -257,6 +267,14 @@
     </{{ element|default('div') }}>
 {%- endblock form_row %}
 
+{% block file_row -%}
+    <{{ element|default('div') }} class="form-group custom-file">
```
Comment (juanmiguelbesada) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: According to [docs](https://getbootstrap.com/docs/4.0/components/forms/#file-browser) and the provided example, form-group class is not needed.

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 2]
```diff
@@ -257,6 +267,14 @@
...
+        {{- form_label(form) -}}
+        {{- form_widget(form) -}}
```
Comment (juanmiguelbesada) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: The label should be after widget

**Review by juanmiguelbesada — CHANGES_REQUESTED** [review]

**Comment by zzenmate** [comment]
I added the most basic block for file_label and file_widget for overriding form-control-label, form-control-file and using custom-file-input, custom-file-label.

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 3]
```diff
@@ -182,7 +182,7 @@
 
 {% block file_widget -%}
     {%- set attr = attr|merge({class: (attr.class|default('') ~ ' custom-file-input')|trim}) -%}
-    {{- block('form_widget_simple') }}
+    {{ block("form_widget_simple", "form_div_layout.html.twig") }}
```
Comment (juanmiguelbesada) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: Double quotes

Comment (juanmiguelbesada) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: why did you remove the "No leading/trailing whitespace"?
Also double quotes should be single quotes

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 4]
```diff
@@ -252,7 +252,7 @@
 
 {% block file_label -%}
     {%- set label_attr = label_attr|merge({class: (label_attr.class|default('') ~ ' custom-file-label')|trim}) -%}
-    {{- block('form_label') -}}
+    {{ block("form_label", "form_div_layout.html.twig") }}
```
Comment (juanmiguelbesada) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: Same here, leading/trailing whitespace and double quotes

**Review by juanmiguelbesada — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Review by juanmiguelbesada — APPROVED** [review]

**Review by javiereguiluz — APPROVED** [review]
It looks good to me, but @Nyholm and @mpiot have more experience with this Bootstrap form theme and could review it too. Thanks!

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 5]
```diff
@@ -180,6 +180,11 @@
     </div>
 {%- endblock choice_widget_expanded %}
 
+{% block file_widget -%}
```
Comment (mpiot) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: In the `form_simple_widget` there is a test on the type of form, if it's a file, it change the class. Adapt it is a better idea than do this block, no ?

    {%- set attr = attr|merge({class: (attr.class|default('') ~ ' form-control' ~ (type|default('') == 'file' ? '-file' : ''))|trim}) -%}

Dot it:

    {%- set attr = attr|merge({class: (attr.class|default('') ~ (type|default('') == 'file' ? 'custom-file-control' : 'form-control'))|trim}) -%}

**Review by mpiot — CHANGES_REQUESTED** [review]

**Comment by mpiot** [comment]
We try to put errors in the label to respect a standard for blind people, but I think it's difficult in this case.. Because the label is directly in the field. I look if there is a way to do it.

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 6]
```diff
@@ -257,6 +267,14 @@
     </{{ element|default('div') }}>
 {%- endblock form_row %}
 
+{% block file_row -%}
+    <{{ element|default('div') }} class="custom-file">
```
Comment (mpiot) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: In your case the element is the div, maybe use <div> directly in place of element ? And you should add a `<div class="input-group mb-3">` or a simple `form-group` or at least the `mb-3` to add a margin after the row.

**Review by mpiot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 6 above)
Comment (juanmiguelbesada) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: Doesn't custom-file-label add a margin-bottom? (.5rem to be exactly)

Comment (mpiot) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: I've try and it's collapsed. In the bootstrap doc, they encapsulate the custom-file-label in an other div with a mb-3 to add the margin bottom.

Comment (Nyholm) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: You need to add a `div` with class "form-group" around this.

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 7]
```diff
@@ -180,6 +180,11 @@
     </div>
 {%- endblock choice_widget_expanded %}
 
+{% block file_widget -%}
+    {%- set attr = attr|merge({class: (attr.class|default('') ~ ' custom-file-input')|trim}) -%}
+    {{- block('form_widget_simple', 'form_div_layout.html.twig') -}}
```
Comment (Nyholm) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: You should remove 'form_div_layout.html.twig'. Or maybe Im missing something. Why did you add it here?

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 5 above)
Comment (Nyholm) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: :+1:

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 8]
```diff
@@ -245,6 +250,11 @@
     {%- endif -%}
 {%- endblock checkbox_radio_label %}
 
+{% block file_label -%}
+    {%- set label_attr = label_attr|merge({class: (label_attr.class|default('') ~ ' custom-file-label')|trim}) -%}
+    {{- block('form_label', 'form_div_layout.html.twig') -}}
```
Comment (Nyholm) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: Do you use `form_div_layout.html.twig` to avoid errors in the label?

**Review by Nyholm — CHANGES_REQUESTED** [review]
Great. I just tested this and it looks okey. I added a few comments.

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 7 above)
Comment (zzenmate) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: I use form_div_layout.html.twig from form_widget_simple for overriding form_widget_simple from bootstrap_4_layout.html.twig.
Because we don't need merge .form-control-file with .custom-file-input.
I think we will need use only one class. And it will be .custom-file-input.
In <a href="https://getbootstrap.com/docs/4.0/components/forms/#file-browser">documention</a> used only .custom-file-input
Or am I wrong?

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 8 above)
Comment (zzenmate) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: As I wrote on top, we don't need to merge 2 class: "custom-file-label" and "form-control-label".
Or am I wrong?

**Comment by zzenmate** [comment]
status: needs review

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 7 above)
Comment (Nyholm) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: Hm... Okey. Why are we adding `form-control-file`. Is that ever needed?

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 8 above)
Comment (Nyholm) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: Im not sure. Ill check later tonight

**Comment by Nyholm** [comment]
There are still some work. Please see https://github.com/symfony/symfony/pull/26502#discussion_r175289610

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 7 above)
Comment (zzenmate) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: https://getbootstrap.com/docs/4.0/components/input-group/#custom-file-input
In this example, we also not used form-control-file or form-control-label.

Comment (Nyholm) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: Excuse my ignorance. But in what scenario/input do we need form-control-file?

Comment (mpiot) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: This block can be removed if you apply the solution specified before: edit the `form_simple_widget`. Your solution remove the way to use the previous possibilities to display file input, the only way become the custom way.

Then, just edit the part where we set the `.form-control-file` in place of `.form-control`, I've write the code to replace on the top. You just need to edit a line in the block `form_simple_widget` and remove this block.

I regive you the line of code:

In the `form_simple_widget` block there is:

    {%- set attr = attr|merge({class: (attr.class|default('') ~ ' form-control' ~ (type|default('') == 'file' ? '-file' : ''))|trim}) -%}

Replace by this line:

    {%- set attr = attr|merge({class: (attr.class|default('') ~ (type|default('') == 'file' ? 'custom-file-input' : 'form-control'))|trim}) -%}

Then, if the type is a file the `attr` become `custom-file-input` in place of `form-control`. The code is already ok, you just have to adapt it.

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 8 above)
Comment (mpiot) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: In this case is it not better to replace (`replace` filter) an existing class in place of do a double override on the block ? Because if we extend a file, that extend a file, each file override a block, and finally we re-override the block again, it's more difficult to understand what we do, no ?

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 9]
```diff
@@ -257,6 +267,16 @@
     </{{ element|default('div') }}>
 {%- endblock form_row %}
 
+{% block file_row -%}
+    <{{ element|default('div') }} class="form-group">
```
Comment (mpiot) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: In this case the element can't be another thing than a `div` element, maybe it's better to directly write `div` in place of use the `element` var ? It works like it, it's just a question/suggestion. ;-)

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 7 above)
Comment (zzenmate) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: We need form-control-file for adding: "display: block; width: 100%;". 
But, I am don't know if we need to use it in this case.

**Review by mpiot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 7 above)
Comment (mpiot) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: In the demo they write it:

    <div class="input-group mb-3">
      <div class="custom-file">
        <input type="file" class="custom-file-input" id="inputGroupFile02">
        <label class="custom-file-label" for="inputGroupFile02">Choose file</label>
      </div>
    </div>

The input just have the class `custom-file-input`, this is what do the line:

    {%- set attr = attr|merge({class: (attr.class|default('') ~ (type|default('') == 'file' ? 'custom-file-input' : 'form-control'))|trim}) -%}

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 8 above)
Comment (zzenmate) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: Yes, you're right. But, How to get input type in form_label for write logic with if.
I mean something like that:
```
{% if type == 'file' %} custom-file-label {% else %} form-control-label {% endif %}
```

Comment (mpiot) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: Yes, in the `block form_label`, you can add this:

    {% elseif type is defined and type == 'file' %}
        {%- set label_attr = label_attr|merge({for: id, class: (label_attr.class|default('') ~ ' custom-file-label')|trim}) -%}

To obtain:

    {%- if compound is defined and compound -%}
        {%- set element = 'legend' -%}
        {%- set label_attr = label_attr|merge({class: (label_attr.class|default('') ~ ' col-form-label')|trim}) -%}
    {% elseif type is defined and type == 'file' %}
        {%- set label_attr = label_attr|merge({for: id, class: (label_attr.class|default('') ~ ' custom-file-label')|trim}) -%}
    {%- else -%}
        {%- set label_attr = label_attr|merge({for: id, class: (label_attr.class|default('') ~ ' form-control-label')|trim}) -%}
    {%- endif -%}

Then, you can remove your the block `file_label`. Keep the `type is defined` test, because it appears that it is not always defined.

Comment (mpiot) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: Ha, and then you need to change something fo errors, because by using our `form_label`, you have the errors displayed in it.

I've try, and if you remove the error handling in your `file_row`, error are displayed semi-correctly... But I think it's ok.

In this case, you can both: use your `file_row`, or add a check on `type == file` in the `form_row`. I think it's more readable with your way, but we re-write code, for it, do as you wish :-) But you need to remove the `form_error` from the `file_row` since you use our `form_label` to avoid error duplication.

Comment (zzenmate) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: @mpiot Thank you for your help.

**Review by mpiot — APPROVED** [review]
It seems all is ok, thanks for your help :-)

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 10]
```diff
@@ -117,7 +117,7 @@
 
 {% block form_widget_simple -%}
     {% if type is not defined or type != 'hidden' %}
-        {%- set attr = attr|merge({class: (attr.class|default('') ~ ' form-control' ~ (type|default('') == 'file' ? '-file' : ''))|trim}) -%}
+        {%- set attr = attr|merge({class: (attr.class|default('') ~ (type|default('') == 'file' ? ' custom-file-input' : ' form-control'))|trim}) -%}
```
Comment (Nyholm) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: You need a space here. 

“ file”

**Review by Nyholm — CHANGES_REQUESTED** [review]
Good. Just a super minor

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 10 above)
Comment (mpiot) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: A space ? It's a test on the `type` of the form, why add a space before ? It can't match if we add it.

Comment (Nyholm) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: No, you are correct. It is the value that should have space and they do.

**Review by Nyholm — APPROVED** [review]
Thank you for this PR. 

im 👍

**Review by Nyholm — APPROVED** [review]
Wrong button :/

Im 👍

**Comment by fabpot** [comment]
Thank you @zzenmate.
